src/valid/handles.rs

```diff
@@ -0,0 +1,337 @@
+                        this_handle.check_dep(image)?.check_dep_opt(match query {
+                            crate::ImageQuery::Size { level } => level,
+                            crate::ImageQuery::NumLevels
+                            | crate::ImageQuery::NumLayers
+                            | crate::ImageQuery::NumSamples => None,
+                        })?;
+                    }
+                    &crate::Expression::Unary {
+                        op: _,
+                        expr: operand,
+                    } => {
+                        this_handle.check_dep(operand)?;
+                    }
+                    &crate::Expression::Binary { op: _, left, right } => {
+                        this_handle.check_dep(left)?.check_dep(right)?;
+                    }
+                    &crate::Expression::Select {
+                        condition,
+                        accept,
+                        reject,
+                    } => {
+                        this_handle
+                            .check_dep(condition)?
+                            .check_dep(accept)?
+                            .check_dep(reject)?;
+                    }
+                    &crate::Expression::Derivative {
+                        axis: _,
+                        expr: argument,
+                    } => {
+                        this_handle.check_dep(argument)?;
+                    }
+                    &crate::Expression::Relational { fun: _, argument } => {
+                        this_handle.check_dep(argument)?;
+                    }
+                    &crate::Expression::Math {
+                        fun: _,
+                        arg,
+                        arg1,
+                        arg2,
+                        arg3,
+                    } => {
+                        this_handle
+                            .check_dep(arg)?
+                            .check_dep_opt(arg1)?
+                            .check_dep_opt(arg2)?
+                            .check_dep_opt(arg3)?;
+                    }
+                    &crate::Expression::As {
+                        expr: input,
+                        kind: _,
+                        convert: _,
+                    } => {
+                        this_handle.check_dep(input)?;
+                    }
+                    &crate::Expression::CallResult(function) => {
+                        function.check_valid_for(functions)?;
+                    }
+                    &crate::Expression::AtomicResult { .. } => (),
+                    &crate::Expression::ArrayLength(array) => {
+                        this_handle.check_dep(array)?;
+                    }
+                }
+            }
+            Ok(())
+        };
+
+        let validate_function = |function: &_| {
+            let &crate::Function {
+                name: _,
+                arguments: _,
+                result: _,
```

Comment:
The arguments and result have type handles.

Reply:
Oh, oof, I feel silly for missing this. 😳 Resolved with 3fba607f05b3f958d45e6139453cbdb5240fd5e0.